### PR TITLE
Upgrade sheltering transfer

### PR DESCRIPTION
**Pull request by AndrewAR2**
# Checklist:

- [x] Code follows the style guidelines
- [x] Documentation updated
- [x] Corner cases evaluated and handled
- [x] index.js exports updated
- [x] Automatic tests added/updated
- [x] Automatic tests pass
- [x] Sanity check passed
- [x] Self-review passed
- [x] Version updated

### CODE REVIEW

File: contracts/Front/ShelteringTransfers.sol [hunk 1]
```diff
@@ -67,20 +81,65 @@ contract ShelteringTransfers is Base {
...
         return donorId != address(0x0);
     }
 
+    function canResolve(address resolverId, bytes32 transferId) public view returns (bool) {
+        (address donorId, bytes32 bundleId, ) = shelteringTransfersStore.getTransfer(transferId);
+
+        // solium-disable-next-line operator-whitespace
+        return donorId != address(0x0) &&
+        !sheltering.isSheltering(bundleId, resolverId) &&
+        resolverId == getTransferDesignatedShelterer(transferId);
```
Comment (vanruch) [inline contracts/Front/ShelteringTransfers.sol]: Why do we want to add DMP into sheltering transfers? The aim of transfers is to get rid of our bundles as soon as possible. Also, I don't understand why introduce `canResolve` when we have `requireResolutionPossible` doing the same thing

File: contracts/Storage/ShelteringTransfersStore.sol [hunk 2]
```diff
@@ -16,25 +16,27 @@ contract ShelteringTransfersStore is Base {
     struct Transfer {
         address donorId;
         bytes32 bundleId;
+        uint64 creationTime;
```
Comment (vanruch) [inline contracts/Storage/ShelteringTransfersStore.sol]: As we don't need DMP, let's not modify state

File: contracts/Front/ShelteringTransfers.sol (hunk 1 above)
Comment (AndrewAR2) [reply contracts/Front/ShelteringTransfers.sol]: In response to TransferStarted event, all Atlases will start downloading the bundle from the donor, which will lead to a DDoS attack on it. Therefore, we need to limit the number of possible resolvers, using DMP as the proven and effective solution.

Comment (AndrewAR2) [reply contracts/Front/ShelteringTransfers.sol]: requireResolutionPossible method has been removed from ShelteringTransfers because its functionality has been moved to canResolve.

**Comment by krzysztof-jelski** [comment]
Currently the contracts upgrade procedure will not upgrade the storage contracts, which means that `ShelteringsTransferStore` won't be deployed. Conceptually it can be deployed, because we have no transfers yet. But we will need to develop an upgrade procedure for this.

**Comment by AndrewAR2** [comment]
Thank you! I will study the contracts upgrade procedure and then offer a solution.

File: contracts/Front/ShelteringTransfers.sol (hunk 1 above)
Comment (krzysztof-jelski) [reply contracts/Front/ShelteringTransfers.sol]: Do we have any data that DMP here is necessary? I'd love to see some data backing up this decision.
I mean we can measure how much load a typical (small DO droplet) Atlas server can handle. We can also estimate the number of requests it will get (expected number of Atlases in the network x bundles the donor wants to transfer).

Comment (AndrewAR2) [reply contracts/Front/ShelteringTransfers.sol]: Suppose that we have 500 Atlases and an average bundle size of 2 MB (in a cryptoeconomy, the average bundle size is 50 MB) - hence each transfer request will generate 1 GB of traffic. And this is only one transfer request! If requests will be 100? I think that such traffic will completely occupy Atlas transmission channel.

**Comment by AndrewAR2** [comment]
I think the easiest way to deploy sheltering transfer contracts is to make temporary changes into method recycleStorageContracts of DeployActions:

for (const contractName of storageContractNames) {
  if (contractName === 'shelteringTransfersStore') {
    continue;
  }
  recycled[contractName] = await this.headWrapper.contractAddressByName(contractName);
}

File: contracts/Front/ShelteringTransfers.sol [hunk 3]
```diff
@@ -67,28 +80,68 @@ contract ShelteringTransfers is Base {
...
+    }
+
+    function getTransferCreationTime(bytes32 transferId) public view returns (uint64) {
+        (,, uint64 creationTime) = shelteringTransfersStore.getTransfer(transferId);
+        return creationTime;
+    }
+
+    function getTransferDesignatedShelterer(bytes32 transferId) public view returns (address) {
```
Comment (krzysztof-jelski) [inline contracts/Front/ShelteringTransfers.sol]: `getTransferDesignatedShelterer ` and `getDesignatedSheltererTier` are duplicates of methods in `Challenges`. Please extract superclass.

Comment (AndrewAR2) [reply contracts/Front/ShelteringTransfers.sol]: getDesignatedSheltererTier - is ok!
But getTransferDesignatedShelterer uses the transfer creation time for round computation and transferId as dmpBaseHash.

File: src/wrappers/sheltering_transfers_wrapper.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+import {MIN_BLOCK_TIME} from '../constants';
+import ManagedContractWrapper from './managed_contract_wrapper';
+
+export default class ShelteringTransfersWrapper extends ManagedContractWrapper {
+  get getContractName() {
+    return 'shelteringTransfers';
+  }
+
+  async earliestMeaningfulBlock(transferDuration) {
```
[the remark this answers is not in the document]
Comment (AndrewAR2) [reply src/wrappers/sheltering_transfers_wrapper.js]: Theoretically, a situation is possible when for TransferStarted there is no corresponding TransferResolved or TransferCancelled. If we do not limit the search for events in the blockchain, then at the start of Atlas we will find these unanswered requests. To prevent this, I propose to limit the search for events in the blockchain to the challenge duration.

[the remark this answers is not in the document]
Comment (krzysztof-jelski) [reply src/wrappers/sheltering_transfers_wrapper.js]: Still I can't see any reason this should be in the wrapper.
The Atlas worker itself can employ any sensible strategy for limiting the search.

File: contracts/Front/ShelteringTransfers.sol (hunk 1 above)
Comment (AndrewAR2) [reply contracts/Front/ShelteringTransfers.sol]: Do you agree that DMP is needed here?

Comment (krzysztof-jelski) [reply contracts/Front/ShelteringTransfers.sol]: Yes, I agree.

File: contracts/Front/DmpBasedDesignator.sol [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    function requestIsInProgress(bytes32 requestId) public view returns (bool);
+
+    function canResolve(address resolverId, bytes32 requestId) public view returns (bool) {
+        bytes32 bundleId = getRequestedBundle(requestId);
+
+        // solium-disable-next-line operator-whitespace
+        return requestIsInProgress(requestId) &&
+        !sheltering.isSheltering(bundleId, resolverId) &&
+        resolverId == getRequestDesignatedShelterer(requestId) &&
+        atlasStakeStore.getStake(resolverId) > 0;
```
Comment (krzysztof-jelski) [inline contracts/Front/DmpBasedDesignator.sol]: Putting `canResolve` changes functionality for `ShelteringTransfers` contract. Specifically, it didn't check `atlasStakeStore.getStake(resolverId) > 0` condition before. I would say it makes sense to make sure the resolver still has stake when he wants to resolve a transfer. So the change itself is ok. But please write an additional test that covers this case for `ShelteringTransfers` contract.

Comment (krzysztof-jelski) [reply contracts/Front/DmpBasedDesignator.sol]: Or maybe we can remove `canResolve` from the superclass, as it's too loosely connected to DMP.

Comment (AndrewAR2) [reply contracts/Front/DmpBasedDesignator.sol]: Remove canResolve from the superclass?
